server/services/ai.py

```diff
@@ -0,0 +1,181 @@
+        """Gets a http session."""
+        retry = urllib3.util.retry.Retry(
+            total=5,
+            status_forcelist=[429, 503],
+            backoff_factor=2,
+            method_whitelist=["POST"],
+        )
+        adapter = requests.adapters.HTTPAdapter(max_retries=retry)
+        session = requests.Session()
+        session.headers.update({
+            "Content-Type": "application/json",
+            "User-Agent": "datcom-inference-client",
+        })
+        session.mount("https://", adapter)
+        return session
+
+    def _get_headers(self) -> Optional[Mapping[str, str]]:
+        if not self.creds:
+            return None
+        if not self.creds.valid:
+            self.creds.refresh(google.auth.transport.requests.Request())
+        return {
+            "X-Goog-User-Project": self.project_id,
+            "Authorization": "Bearer " + self.creds.token,
+        }
+
+    def request(self, query: str):
+        headers = self._get_headers()
+        json_data = json.dumps({"instances": [query]})
+        return self.session.post(self.url,
+                                 data=json_data,
+                                 headers=headers,
+                                 timeout=self.predict_timeout).json()
+
+
+def get_region() -> Optional[str]:
+    """Returns a string indicating the region."""
+    metadata_url = "http://metadata.google.internal/computeMetadata/v1/instance/zone"
+    metadata_flavor = {"Metadata-Flavor": "Google"}
+    zone = requests.get(metadata_url, headers=metadata_flavor).text
+    # zone is in the format of projects/projectnum/zones/zone
+    region = "-".join(zone.split("/")[3].split("-")[0:2])
+    return region
+
+
+def create_inference_client(config_path: str) -> Optional[InferenceClient]:
+    """Creates an InferenceClient instance for generic model predictions."""
+
+    logging.info("Loading InferenceClient using config file %s", config_path)
+
+    factory = {
+        "grpc": GrpcInferenceClient,
+        "rest": RestInferenceClient,
+    }
+
+    with open(config_path) as f:
+        ai = yaml.full_load(f)
+        # We use a fake region when locally testing with LocalConfig.
+        region = ai["local"]["region"] if cfg.LOCAL else get_region()
+        if region in ai:
+            endpoint_id = ai[region]["endpoint_id"]
+            deployed_model_id = ai[region].get("deployed_model_id", "")
+            protocol = ai[region]["protocol"]
+            cls = factory.get(protocol, None)
+            if cls:
+                return cls(region, endpoint_id, deployed_model_id)
+            raise ValueError(f"Unknown protocol {protocol}")
+
+
+if not cfg.TEST and cfg.AI_CONFIG_PATH:
+    _INFERENCE_CLIENT = create_inference_client(cfg.AI_CONFIG_PATH)
+
+
+def search(query: str) -> Optional[Any]:
+    global _INFERENCE_CLIENT
+
+    if _INFERENCE_CLIENT:
+        response = _INFERENCE_CLIENT.request(query)
+        # This is a fake response for the moment.
```

Comment:
To confirm, does this return real inference for dev instance?

Reply:
Yes. I tested it using the public endpoint. But I also like to test it on a dev environment to make sure the grpc client works as expected.

Do you have any instructions on how to do that?

Reply:
Tested on dev. Seems to work fine.